galley/pkg/config/analysis/analyzers/gateway/conflictinggateway.go

```diff
@@ -0,0 +1,151 @@
+	"istio.io/istio/galley/pkg/config/analysis/msg"
+	"istio.io/istio/pkg/config/host"
+	"istio.io/istio/pkg/config/resource"
+	"istio.io/istio/pkg/config/schema/collection"
+	"istio.io/istio/pkg/config/schema/collections"
+)
+
+// ConflictingGatewayAnalyzer checks a gateway's selector, port number and hosts.
+type ConflictingGatewayAnalyzer struct{}
+
+// (compile-time check that we implement the interface)
+var _ analysis.Analyzer = &ConflictingGatewayAnalyzer{}
+
+// Metadata implements analysis.Analyzer
+func (*ConflictingGatewayAnalyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name:        "gateway.ConflictingGatewayAnalyzer",
+		Description: "Checks a gateway's selector, port number and hosts",
+		Inputs: collection.Names{
+			collections.IstioNetworkingV1Alpha3Gateways.Name(),
+		},
+	}
+}
+
+// Analyze implements analysis.Analyzer
+func (s *ConflictingGatewayAnalyzer) Analyze(c analysis.Context) {
+	gwConflictingMap := initGatewaysMap(c)
+	c.ForEach(collections.IstioNetworkingV1Alpha3Gateways.Name(), func(r *resource.Instance) bool {
+		s.analyzeGateway(r, c, gwConflictingMap)
+		return true
+	})
+}
+
+func (*ConflictingGatewayAnalyzer) analyzeGateway(r *resource.Instance, c analysis.Context,
+	gwCMap map[string]map[string][]string) {
+	gw := r.Message.(*v1alpha3.Gateway)
+	gwName := r.Metadata.FullName.String()
+	// For pods selected by gw.Selector, find Services that select them and remember those ports
+	gwSelector := k8s_labels.SelectorFromSet(gw.Selector)
+	sGWSelector := gwSelector.String()
+
+	// Check non-exist gateway with particular selector
+	var isExists bool = false
+	for gwmKey := range gwCMap {
+		if strings.Contains(gwmKey, sGWSelector) {
+			isExists = true
+			break
+		}
+	}
+	if !isExists {
+		m := msg.NewReferencedResourceNotFound(r, "selector", sGWSelector)
+		label := util.ExtractLabelFromSelectorString(sGWSelector)
+		if line, ok := util.ErrorLine(r, fmt.Sprintf(util.GatewaySelector, label)); ok {
+			m.Line = line
+		}
+		c.Report(collections.IstioNetworkingV1Alpha3Gateways.Name(), m)
+		return
+	}
+
+	var rmsg []string
+	conflictingGWMatch := 0
+	for _, server := range gw.Servers {
+		sPortNumber := strconv.Itoa(int(server.Port.Number))
+		mapKey := GenGatewayMapKey(sGWSelector, sPortNumber)
+		for gwNameKey, gwHostsValue := range gwCMap[mapKey] {
+			for _, gwHost := range server.Hosts {
+				// both selector and portnumber are the same, then check hosts
+				if IsGWsHostMatched(gwHost, gwHostsValue) {
+					if gwName != gwNameKey {
+						conflictingGWMatch++
+						rmsg = append(rmsg, gwNameKey)
+					}
+				}
+			}
+		}
+	}
+	if conflictingGWMatch > 0 {
+		reportMsg := fmt.Sprintf("%s to %s", gwName, strings.Join(rmsg, ","))
+		m := msg.NewConflictingGateways(r, reportMsg)
+		c.Report(collections.IstioNetworkingV1Alpha3Gateways.Name(), m)
+	}
+}
+
+// IsGWsHostMatched implements gateway's hosts match
+func IsGWsHostMatched(gwInstance string, gwHostList []string) bool {
```

Comment:
Let's keep this func private if it's not going to be used outside this analyzer.

Reply:
okay, I will make it invisibility outside of this analyzer